### PR TITLE
[WIP][JENKINS-27648] Introduce resource disposal service

### PR DESCRIPTION
**Pull request by olivergondza**
This is a proof of concept implementation of [JENKINS-27648](https://issues.jenkins-ci.org/browse/JENKINS-27648), service that asynchronously disposes registered resources.

This change makes `FilePath#deleteRecursive()` asynchronous operation.

@reviewbybees

### CODE REVIEW

**Comment by jglick** [comment]
> This change makes `FilePath#deleteRecursive()` asynchronous operation.

-1 on that at least. Incompatible and unexpected.

**Comment by olivergondza** [comment]
@jglick, is that incompatible if we rename it to random sibling dir and then delete? The directory name can be reused as soon as the call completes. Just the disk space is not freed immediately (which is not guaranteed on all file systems anyway).

**Comment by jglick** [comment]
> is that incompatible if we rename it to random sibling dir and then delete?

Probably harmless. Might be a few tests which get broken (say, if they are calling `deletedDir.getParent().list()`).

More broadly I am concerned that this is making the system more complex without a clear reason I can see. Deletion may indeed fail under some circumstances but these usually involve inappropriate file locks being held, which is corrected by closing streams reliably and so on. And deletion being slow is only a problem if the UI is blocked waiting for deletion to finish, which is better solved by providing a more responsive UI that does not call I/O operations from the request handler thread.

**Comment by olivergondza** [comment]
The IO operations can be [offloaded](https://issues.jenkins-ci.org/browse/JENKINS-24824) from critical paths. Though, there is no nice way to retry and report problems to admins. Also, my intention is to replace https://github.com/jenkinsci/jenkins/pull/1209.

File: core/src/main/java/hudson/FilePath.java [hunk 1]
```diff
@@ -1166,13 +1175,60 @@ public Boolean invoke(File f, VirtualChannel channel) throws IOException, Interr
...
+                // Failed to rename or schedule remotely
             }
-        });
+
+            // Fallback to synchronous delete when not on master or not talking to master
+            if (renamed) {
+                tempFile.renameTo(file);
+            }
+
+            file.act(this);
```
Comment (jtnord) [inline core/src/main/java/hudson/FilePath.java]: so if the rename worked on windows you can delete the directory recursivly - so this just adds unnessecary overhead.
If you can't rename the directory on windows your not going to be able to delete it either.

**Comment by jtnord** [comment]
If you can't delete a directory on windows because a File is open inside it - you can't rename the parent folder either.   So This seems DOA for the intended use case.

I would much rather that issues are fixed in the plugins - as doing this asynchronously is just a PITA for unit tests.

so a general -1 on the approach.

**Comment by olivergondza** [comment]
> If you can't delete a directory on windows because a File is open inside it - you can't rename the parent folder either

Was not aware of that.

**Comment by olivergondza** [comment]
Thanks for the review. Back to the drawing board...

**Pull request closed without merging**